File: android/data/repositories/src/main/kotlin/com/gemwallet/android/data/repositories/support/SupportTypingState.kt
package com.gemwallet.android.data.repositories.support

import com.wallet.core.primitives.SupportAgent
import com.wallet.core.primitives.SupportTyping
import com.wallet.core.primitives.SupportTypingStatus
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class SupportTypingState {

    private val state = MutableStateFlow<SupportAgent?>(null)
    val agent: StateFlow<SupportAgent?> = state.asStateFlow()

    fun update(typing: SupportTyping) {
        state.value = when (typing.status) {
            SupportTypingStatus.On -> typing.agent
            SupportTypingStatus.Off -> null
        }
    }

    fun clear() {
        state.value = null
    }
}

File: ios/Packages/FeatureServices/SupportChatService/SupportTypingState.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import Primitives

@Observable
public final class SupportTypingState: Sendable {
    @MainActor public private(set) var agent: SupportAgent?

    public init() {}

    @MainActor
    public func update(_ typing: SupportTyping) {
        switch typing.status {
        case .on:
            agent = typing.agent
        case .off:
            clear()
        }
    }

    @MainActor
    public func clear() {
        agent = nil
    }
}
